components/support/rc_crypto/nss/src/pk11/types.rs

```diff
@@ -0,0 +1,213 @@
+            type_: nss_sys::SECItemType::siBuffer,
+            data: obj_id_buf.as_ptr() as *mut c_uchar,
+            len: c_uint::try_from(obj_id_buf.len())?,
+        };
+        let slot = get_internal_slot()?;
+        let mut pre_existing_key = unsafe {
+            nss_sys::PK11_FindKeyByKeyID(slot.as_mut_ptr(), &mut obj_id, std::ptr::null_mut())
+        };
+        if !pre_existing_key.is_null() {
+            // Note that we can't just call SECKEY_DestroyPrivateKey here because that
+            // will destroy the PKCS#11 object that is backing a preexisting key (that
+            // we still have a handle on somewhere else in memory). If that object were
+            // destroyed, cryptographic operations performed by that other key would
+            // fail.
+            unsafe {
+                destroy_private_key_without_destroying_pkcs11_object(pre_existing_key);
+            }
+            // Try again with a new ID (but only once - collisions are very unlikely).
+            generate_random(&mut obj_id_buf)?;
+            pre_existing_key = unsafe {
+                nss_sys::PK11_FindKeyByKeyID(slot.as_mut_ptr(), &mut obj_id, std::ptr::null_mut())
+            };
+            if !pre_existing_key.is_null() {
+                unsafe {
+                    destroy_private_key_without_destroying_pkcs11_object(pre_existing_key);
+                }
+                return Err(ErrorKind::InternalError.into());
+            }
+        }
+        let template_len = c_int::try_from(template.len())?;
+        let mut id_attr: &mut nss_sys::CK_ATTRIBUTE = template
+            .iter_mut()
+            .find(|&&mut attr| {
+                attr.type_ == nss_sys::CKA_ID.into()
+                    && attr.pValue.is_null()
+                    && attr.ulValueLen == 0
+            })
+            .ok_or_else(|| ErrorKind::InternalError)?;
+        id_attr.pValue = obj_id_buf.as_mut_ptr() as *mut c_void;
+        id_attr.ulValueLen = nss_sys::CK_ULONG::try_from(obj_id_buf.len())?;
+        // We use `PK11_CreateGenericObject` instead of `PK11_CreateManagedGenericObject`
+        // to leak the reference on purpose because `PK11_FindKeyByKeyID` will take
+        // ownership of it.
+        let _obj = unsafe {
+            GenericObject::from_ptr(nss_sys::PK11_CreateGenericObject(
+                slot.as_mut_ptr(),
+                template.as_mut_ptr(),
+                template_len,
+                nss_sys::PR_FALSE,
+            ))?
+        };
+        // Have NSS translate the object to a private key.
+        Ok(unsafe {
+            PrivateKey::from_ptr(nss_sys::PK11_FindKeyByKeyID(
+                slot.as_mut_ptr(),
+                &mut obj_id,
+                std::ptr::null_mut(),
+            ))?
+        })
+    }
+}
+
+// This is typically used by functions receiving a pointer to an `out SECItem`,
+// where we allocate the struct, but NSS allocates the elements it points to.
+pub(crate) struct ScopedSECItem {
```

Comment:
This one makes me sweat still... But not a lot we can do.

Reply:
I guess you left comments on this in the other PR, but would you mind summarizing your concerns here for posterity @thomcc?